ql/src/java/org/apache/hadoop/hive/ql/optimizer/ColumnPrunerProcCtx.java

```diff
@@ -298,26 +302,130 @@ public ParseContext getParseContext() {
+  private List<String> getNestedColPathByExpr(ExprNodeDesc expr) {
+    List<String> res = new ArrayList<>();
+    //TODO do we need to check isList here?
     if (expr instanceof ExprNodeFieldDesc && !((ExprNodeFieldDesc) expr).getIsList()) {
       ExprNodeFieldDesc exprNodeFieldDesc = (ExprNodeFieldDesc) expr;
-      List<String> cols = exprNodeFieldDesc.getCols();
-      if (cols != null && !cols.isEmpty()) {
-        return StringUtils
-          .join(".", new String[] { cols.get(0), exprNodeFieldDesc.getFieldName() });
+      ExprNodeDesc desc = exprNodeFieldDesc.getDesc();
+      TypeInfo typeInfo = desc.getTypeInfo();
+      final TypeInfo fieldTypeInfo = exprNodeFieldDesc.getTypeInfo();
+      String fieldName = exprNodeFieldDesc.getFieldName();
+      if (fieldName == null || fieldName.isEmpty()) {
+        return res;
+      }
+      switch (typeInfo.getCategory()) {
+      case PRIMITIVE:
+        // Do nothing since it's not a nested column attribution
+        break;
+      case LIST:
+        //TODO
+        break;
+      case MAP:
+        //TODO
+        break;
+      case STRUCT:
+        // Not a list so the column name is the first item
+        if (fieldTypeInfo.equals(typeInfo) && fieldName.equals(desc.getCols().get(0))) {
+          return res;
+        } else {
+          res.addAll(
+            findAllLeavesByField((StructTypeInfo) typeInfo, desc.getCols().get(0), fieldName,
+              fieldTypeInfo));
+        }
+        break;
+      case UNION:
+        //TODO
+        break;
+      default:
+        throw new RuntimeException("Unknown type founded");
+      }
+    }
+    return res;
+  }
+
+  private List<String> findAllLeavesByField(
+    StructTypeInfo structTypeInfo,
+    String pathToRoot,
+    String fieldName,
+    TypeInfo fieldTypeInfo) {
+    List<String> res = new ArrayList<>();
+    List<TypeInfo> types = structTypeInfo.getAllStructFieldTypeInfos();
+    List<String> typeNames = structTypeInfo.getAllStructFieldNames();
+    for (int i = 0; i < typeNames.size(); i++) {
+      String n = typeNames.get(i);
+      String currentPath = pathToRoot.isEmpty() ? n : pathToRoot + "." + n;
+      TypeInfo t = types.get(i);
+      switch (t.getCategory()) {
+      case PRIMITIVE:
+        if(n.equals(fieldName) && t.equals(fieldTypeInfo)){
+          res.add(currentPath);
+        }
+        break;
+      case LIST:
+        //TODO
+        break;
+      case MAP:
+        //TODO
+        break;
+      case STRUCT:
+        if (n.equals(fieldName) && t.equals(fieldTypeInfo)) {
+          res.addAll(getAllLeavesPaths((StructTypeInfo) t, currentPath));
+        } else {
+          res.addAll(findAllLeavesByField((StructTypeInfo) t, currentPath, fieldName, fieldTypeInfo));
+        }
+        break;
+      case UNION:
+        //TODO
+        break;
+      default:
+        throw new RuntimeException("Unknown type founded");
+      }
+    }
+    return res;
+  }
+
+  private List<String> getAllLeavesPaths(
```

Comment:
Perhaps this function is not necessary? we can just add the logic in line 383 above.